pkg/vm/vm_core.go

```diff
@@ -1,15 +1,173 @@
+import (
+	"errors"
+	"fmt"
+
+	"github.com/lambdaclass/cairo-vm.go/pkg/vm/memory"
+)
+
+type VirtualMachineError struct {
+	Msg string
+}
+
+func (e *VirtualMachineError) Error() string {
+	return fmt.Sprintf(e.Msg)
+}
 
 // VirtualMachine represents the Cairo VM.
 // Runs Cairo assembly and produces an execution trace.
 type VirtualMachine struct {
 	RunContext  RunContext
-	currentStep uint
+	CurrentStep uint
 	Segments    memory.MemorySegmentManager
 }
 
-func NewVirtualMachine() *VirtualMachine {
-	return &VirtualMachine{Segments: *memory.NewMemorySegmentManager()}
+func NewVirtualMachine() VirtualMachine {
+	return VirtualMachine{Segments: memory.NewMemorySegmentManager()}
+}
+
+type Operands struct {
+	Dst memory.MaybeRelocatable
+	Res *memory.MaybeRelocatable
+	Op0 memory.MaybeRelocatable
+	Op1 memory.MaybeRelocatable
+}
+
+type OperandsAddresses struct {
+	Dst_addr memory.Relocatable
+	Op0_addr memory.Relocatable
+	Op1_addr memory.Relocatable
+}
+
+// ------------------------
+//  Deduced Operands funcs
+// ------------------------
+
+type DeducedOperands struct {
+	Operands uint8
+}
+
+func (vm *VirtualMachine) OpcodeAssertions(instruction Instruction, operands Operands) error {
+	switch instruction.Opcode {
+	case AssertEq:
+		if operands.Res == nil {
+			return &VirtualMachineError{"UnconstrainedResAssertEq"}
+		}
+		if !operands.Res.IsEqual(&operands.Dst) {
+			return &VirtualMachineError{"DiffAssertValues"}
+		}
+	case Call:
+		new_rel, err := vm.RunContext.Pc.AddUint(instruction.size())
+		if err != nil {
+			return err
+		}
+		returnPC := memory.NewMaybeRelocatableRelocatable(new_rel)
+
+		if !operands.Op0.IsEqual(returnPC) {
+			return &VirtualMachineError{"CantWriteReturnPc"}
+		}
+
+		returnFP := vm.RunContext.Fp
+		dstRelocatable, _ := operands.Dst.GetRelocatable()
+		if !returnFP.IsEqual(&dstRelocatable) {
+			return &VirtualMachineError{"CantWriteReturnFp"}
+		}
+	}
+
+	return nil
+}
+
+func (deduced *DeducedOperands) set_dst(value uint8) {
+	deduced.Operands = deduced.Operands | value
+}
+
+func (deduced *DeducedOperands) set_op0(value uint8) {
+	deduced.Operands = deduced.Operands | value<<1
+}
+
+func (deduced *DeducedOperands) set_op1(value uint8) {
+	deduced.Operands = deduced.Operands | value<<2
+}
+
+//------------------------
+//  virtual machines funcs
+// ------------------------
+
+func VmNew(run_context RunContext, current_step uint, segments_manager memory.MemorySegmentManager) VirtualMachine {
+	return VirtualMachine{
+		RunContext:  run_context,
+		CurrentStep: current_step,
+		Segments:    segments_manager,
+	}
+}
+
+func (vm *VirtualMachine) ComputeRes(instruction Instruction, op0 memory.MaybeRelocatable, op1 memory.MaybeRelocatable) (memory.MaybeRelocatable, error) {
+	switch instruction.ResLogic {
+	case ResOp1:
+		return op1, nil
+
+	case ResAdd:
+		maybe_rel, err := op0.AddMaybeRelocatable(op1)
+		if err != nil {
+			return memory.MaybeRelocatable{}, errors.New("adding maybe relocatable")
```

Comment:
Shouldn't we just return the err from `AddMaybeRelocatable`?, this message doesn't add any extra information

Reply:
right